AppController/djinn.rb

```diff
@@ -5040,78 +5022,123 @@ def get_scaling_info_for_version(version_key)
-                       "for #{version_key}.")
-        return 0
-      end
-      appservers_to_scale = calculate_appservers_needed(
-          num_appservers, current_sessions, allow_concurrency)
+    # Services need will override the applicaiton desires, in order to
+    # avoid overloading.
+    delta = check_services_load(version_key, num_appservers)
+    appservers_to_scale = delta if delta < 0
 
-      # Let's make sure we don't get over the user define maximum.
-      if num_appservers + appservers_to_scale > max
-        appservers_to_scale = max - num_appservers
-      end
+    # Gets the scaling limits for this application.
+    min, max = get_min_max_appservers(version_details)
 
-      Djinn.log_debug("The deployment has reached its maximum load " \
-                      "threshold for #{version_key} - Advising that we " \
-                      "scale up #{appservers_to_scale} AppServers.")
-      return appservers_to_scale
+    # Make sure we don't go over the maximum number of AppServers
+    # indicated for this application.
+    if num_appservers + appservers_to_scale > max
+      appservers_to_scale = max - num_appservers
+      if appservers_to_scale.zero?
+        Djinn.log_debug("#{version_key} has reached the maximum number of allowed " \
+                        "AppServers (#{max}).")
+      end
+    elsif num_appservers + appservers_to_scale < min
+      appservers_to_scale = num_appservers - min
+      if appservers_to_scale.zero?
+        Djinn.log_debug("#{version_key} has reached the minimum number of allowed " \
+                        "AppServers (#{max}).")
+      end
+    end
 
-    elsif current_load <= MIN_LOAD_THRESHOLD
+    # We have a cooldown period for downscaling to ensure we give enough
+    # time to the stats to be representative before removing too many
+    # AppServers.
+    if appservers_to_scale < 0
       downscale_cooldown = SCALEDOWN_THRESHOLD * DUTY_CYCLE
       if Time.now.to_i - @last_decision[version_key] < downscale_cooldown
         Djinn.log_debug(
           "Not enough time has passed to scale down #{version_key}")
-        return 0
+        appservers_to_scale = 0
       end
-      appservers_to_scale = calculate_appservers_needed(
-          num_appservers, current_sessions, allow_concurrency)
-      Djinn.log_debug("The deployment is below its minimum load threshold " \
-                      "for #{version_key} - Advising that we scale down " \
-                      "#{appservers_to_scale.abs} AppServers.")
-      return appservers_to_scale
-    else
-      Djinn.log_debug("The deployment is within the desired range of load " \
-                      "for #{version_key} - Advising that there is no need " \
-                      "to scale currently.")
-      return 0
     end
+
+    # Finally we have to ensure we have at least the minimum number of
+    # AppServers required.
+    if num_appservers + appservers_to_scale < min
```

Comment:
Yes they are very similar. I like to have them both because they are inexpensive and the former checks that the application will not scale below the minimum, while this one ensures that the applications has a minimum sets of AppServers. I can remove the 5112 if you like it better, but I prefer to have the last check to be the insurance the application will stay up.